File: app/socket.py
from flask_socketio import SocketIO, emit, join_room, leave_room, close_room
from flask_login import current_user
from .models import db, ChannelMessage, User, Member, Server
import os
import datetime


# configure cors_allowed_origins
if os.environ.get("FLASK_ENV") == "production":
    origins = [
        "https://accord-ajr.onrender.com",
        "https://accord-ajr.onrender.com",
    ]
else:
    origins = "*"

# initialize your socket instance
socketio = SocketIO(cors_allowed_origins=origins)

# Might use for bonus. Make isOnline key for Users table, or make a reducer for sockets. Have to set up Disconnect and Connect listeners.
online_users = set()


# handle connecting user to all servers(rooms)
@socketio.on("join_server")
def join_server(data):
    print("**************************JOIN SERVER DATA START**************************")
    print("online_users before: ", online_users)
    user = User.query.get(data["user_id"])
    servers = user.user_servers()
    online_users.add(user.id)
    # connect user to rooms based on server id
    for server in servers:
        print(f"Joining Server: {server}")
        join_room(str(server))
    print("**************************JOIN SERVER DATA END**************************")
    emit(
        "join_server_response",
        {
            "Message": f"Joined Servers: {servers}",
            "Users": f"Online Users: {online_users}",
        },
        broadcast=True,
    )


@socketio.on("add_server")
def add_server(data):
    print("**************************ADD SERVER DATA START**************************")
    join_room(str(data["server_id"]))
    print("**************************ADD SERVER DATA END**************************")
    emit(
        "join_server_response",
        {
            "Message": f"Joined Server: {data['server_id']}",
            "Users": f"Online Users: {online_users}",
        },
        broadcast=False,
    )


# handle chat messages when message_update event is emitted from the frontend
@socketio.on("chat_update")
def handle_chat(data):
    print("**************************SEND_MESSAGE DATA START**************************")
    print(data)
    print("**************************SEND_MESSAGE DATA END**************************")
    emit(
        "chat_update_response",
        {
            "channel_id": data["channel_id"],
            "server_id": data["server_id"],
            "message_id": data.get("message_id", None),
            "Action_Type": data["action_type"],
            "message": data.get("message", None),
        },
        room=str(data["server_id"]),
        broadcast=False,
    )


@socketio.on("channel_update")
def handle_channel(data):
    print("**************************SEND_MESSAGE DATA START**************************")
    print(data)
    print("**************************SEND_MESSAGE DATA END**************************")
    emit(
        "channel_update_response",
        {
            "channel_id": data["channel_id"],
            "server_id": data["server_id"],
            "Action_Type": data["action_type"],
            "channel_name": data.get("channel_name", None),
        },
        room=str(data["server_id"]),
        broadcast=False,
    )


@socketio.on("disconnect")
def handle_disconnect():
    emit(
        "disconnect_response",
        {
            "Message": "USER HAS DISCONNECTED",
            "Users": f"Online Users: {online_users}",
        },
        broadcast=True,
    )

File: react-app/src/components/utils/Socket.js
import { io } from "socket.io-client";

const socket = io();

/* TODO
-implement chatUpdate to Chat.js (chatUpdate accounts for all CRUD)
  --Update reducers to corresponding thunks
---BONUS---
-create channelUpdate for realtime CRUD
  --Only when viewing a server
-create serverUpdate for realtime CRUD
  --Global
-create memberUpdate for realtime CRUD
  --Only when viewing a server
-setup online/offline feature
  --On login establish all listeners for server wide use
  --On logout disconnect sockets
*/

/*
If BE emits "update_chat_response", send channelID and serverID to FE, Update listener for "update_chanel_response" logic for handle messages
dispatches to only dispatch get messages if the param channel id == channelID send from the backend. If so, dispatch, if not, send data update to other state or give that channel/server an alert icon.
Implement has read or view emitter/listener? if that happens.
*/

/*
Before working withs server(s) reducer, try update thunks to work with current socket set up,
If not, check send message from client to backend for other clients to recieve what type of update to dispatch
*/

export function startListeners(user) {
  console.log("***LISTENING FOR SEND MESSAGE RESPONSE***");
  socket.on("disconnect_response", (data) => {
    console.log(data["Message"]);
    console.log(data["Users"]);
  });
}

export function disconnectSockets() {
  socket.disconnect();
}

export function joinServer(user_id) {
  console.log("***EMIT JOIN SERVER***");
  socket.emit("join_server", { user_id });
  console.log("***LISTENING FOR JOIN SERVER RESPONSE***");
  socket.on("join_server_response", (data) => {
    console.log(data["Message"]);
    console.log(data["Users"]);
  });
}

export function addServer(server_id) {
  console.log("***EMIT ADD SERVER***");
  socket.emit("add_server", { server_id: server_id });
}

export function chatUpdate(payload) {
  socket.emit("chat_update", payload);
}

export function handleChatUpdates(callbacks, chid) {
  socket.on("chat_update_response", (data) => {
    const {
      server_id,
      channel_id,
      message_id,
      Action_Type: actionType,
      message: message,
    } = data;
    if (channel_id == chid && callbacks[actionType]) {
      callbacks[actionType](data);
    }
  });
}

export function channelUpdate(payload) {
  console.log("***EMIT CHANNEL UPDATE***");
  socket.emit("channel_update", payload);
}

export function handleChannelUpdates(callbacks, chid) {
  console.log("***LISTENING FOR CHANNEL UPDATES***");

  socket.on("channel_update_response", (data) => {
    const {
      server_id,
      channel_id,
      Action_Type: actionType,
      channel: channel,
      channel_name: channel_name,
    } = data;
    if (callbacks[actionType]) {
      console.log(`${actionType} CHANNEL SOCKET EMITTED`);
      callbacks[actionType](data);
    }
  });
}

export function memberUpdate(payload) {
  console.log("***EMIT MEMBER UPDATE***");
  console.log(payload);
  socket.emit("member_update", payload);
}

export function handleMemberUpdates(callbacks, chid) {
  console.log("***LISTENING FOR MEMBER UPDATES***");

  socket.on("member_update_response", (data) => {
    const {
      server_id,
      member_id: member_id,
      Action_Type: actionType,
      member: member,
    } = data;
    if (callbacks[actionType]) {
      console.log(`${actionType} MEMBER SOCKET EMITTED`);
      callbacks[actionType](data);
    }
  });
}

export default socket;
